### PR TITLE
feature(i18n): Adds keyword argument support for translations.

### PR DESCRIPTION
**Pull request by brettp**
- [ ] Add more tests for mixing string and int arrays.
- [ ] Test performance.
- [ ] Decide on allowed chars for keywords (ideally match ICU standards)
- [ ] Add escaping

### CODE REVIEW

**Comment by ewinslow** [comment]
FWIW, I tried something similar in the past using a widely standardized syntax:

http://php.net/manual/en/class.messageformatter.php

Can't find the pr on my phone ATM but might be worth looking up.

**Comment by brettp** [comment]
Here's Evan's PR: https://github.com/Elgg/Elgg/pull/8023

**Comment by brettp** [comment]
A summary of the reasons the PR wasn't included:
* It requires a non-default PHP extension `ext-intl`
* Lack of good PHP documentation. 
* Steve wasn't comfortable including an MVP polyfill.

File: engine/classes/Elgg/I18n/Translator.php [hunk 1]
```diff
@@ -18,6 +18,15 @@ class Translator {
 	private $CONFIG;
 
 	/**
+	 * Regexp for finding keyword arguments in language strings.
+	 * Accepts forms like {label}, { label}, and {label } where label can include
+	 * a-z, A-Z, 0-9, _, -, :, /, and .
```
Comment (mrclay) [inline engine/classes/Elgg/I18n/Translator.php]: What's the need for these: `- : / .`?

**Comment by mrclay** [comment]
Ideally should be a way to escape an opening brace so it's never matched.

It's a little odd that arguments are optional, but if none are supplied, none of the keywords are replaced. Definitely helps performance though.

**Comment by mrclay** [comment]
How would the ICU approach impact our use of Transifex?

**Comment by hypeJunction** [comment]
I think we could use a few more test cases:
- multiple identical {label}s
- malformatted {label%s}, {\label}
- escaped characters/brackets
- ascii codes/html entities
- html tag attribute value replacement
- numeric labels with and without keyword args (elgg_echo ('eat {0}', ['more beef']);)
-

**Comment by mrclay** [comment]
@hypeJunction You can already use positionals in sprintf formats: `%2$s` outputs 2nd given value.

We could easily support replacing `{1}`, etc, but it would mean processing *every* translation with preg_replace_callback as well as sprintf. I'd rather not.

**Comment by hypeJunction** [comment]
I don't say we have to support it. It's just not clear how the api is
expected to behave in these cases. What if you pass an array with 0 and '0'
as keys?

**Comment by mrclay** [comment]
PHP auto-casts "int looking" array keys to int, so that's not a problem, but it's unclear what happens if int and string keys are mixed.

File: engine/classes/Elgg/I18n/Translator.php (hunk 1 above)
Comment (brettp) [inline engine/classes/Elgg/I18n/Translator.php]: Flexibility. Our project was using a mix of separating chars: `notifications:group-warning`, `notifications/groups/info` If we're still wanting to roll our version and not adopt ext-intl, we can probably shorten this to "everything that isn't a closing tag."

**Comment by brettp** [comment]
Agreed with the need for more test cases mixing ints and strings.
We should also measure the performance impact. I'm guessing `preg_replace` will be slower than the extension.

If we're not ready to throw the switch on polyfilling or requiring the ext for 3.0, the current implementation is compatible with ICU, so we could introduce that in a later release.

File: engine/classes/Elgg/I18n/Translator.php (hunk 1 above)
Comment (brettp) [inline engine/classes/Elgg/I18n/Translator.php]: Though thinking about it, we were using it for a very specific system that used namespaced _keywords_. That's probably not a common use case, so this can just be `:alnum:`.

**Comment by brettp** [comment]
This is java's implementation of escaping ICU 😱:

```
$out = "'{0}'";
$in = "'''{'0}''";
// or
$in = "'''{0}'''";
```
http://docs.oracle.com/javase/6/docs/api/java/text/MessageFormat.html

**Comment by mrclay** [comment]
Looks like [Symfony's Intl implementation](https://github.com/symfony/symfony/tree/master/src/Symfony/Component/Intl#intl-component) doesn't have MessageFormatter, and only provides the "en" locale. Symfony uses a proprietary translator.

**Comment by brettp** [comment]
I'd like to get this in for 3.0. What does everyone think about pulling in the simple `{keyword}` replacement using the approach we have here? It's BC with MessageFormatter, so keeps us open for integrating full support via intl-ext sometime in the future.

**Comment by mrclay** [comment]
:+1 for current impl, limited to `:alnum:`. I'd like to make [quoting](http://userguide.icu-project.org/formatparse/messages#TOC-Quoting-Escaping) (finally found decent docs) forward compatible.

**Comment by hypeJunction** [comment]
@brettp Any chance we would see this in 3.0? If not, let's close and reopen when you ready.

**Comment by jdalsem** [comment]
Closing this for lack of activity. Reopen if you are ready to work on it again
